driver/utils/unicode_conv.h

```diff
@@ -182,113 +196,84 @@ inline decltype(auto) toUTF8(const std::basic_string<CharType> & src) {
 }
+#endif
 
 template <>
-inline decltype(auto) fromUTF8<unsigned char>(const std::string & src) {
-    const auto converted = fromUTF8<char>(src);
-    return std::basic_string<unsigned char>{converted.begin(), converted.end()};
+inline decltype(auto) fromUTF8<wchar_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_wchar.from_bytes(src);
 }
 
+#if !defined(_MSC_VER) || _MSC_VER >= 1920
 template <>
-inline decltype(auto) fromUTF8<unsigned short>(const std::string & src) {
+inline decltype(auto) fromUTF8<unsigned short>(const std::string & src, UnicodeConversionContext & context) {
     static_assert(sizeof(unsigned short) == sizeof(char16_t), "unsigned short doesn't match char16_t exactly");
-    const auto converted = fromUTF8<char16_t>(src);
+    const auto converted = fromUTF8<char16_t>(src, context);
     return std::basic_string<unsigned short>{converted.begin(), converted.end()};
 }
+#endif
 
 template <typename CharType>
-inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest, const std::locale& locale); // Leave unimplemented for general case.
-
-template <>
-inline void fromUTF8<char>(const std::string & src, std::string & dest, const std::locale& locale) {
-    dest = fromUTF8<char>(src, locale);
-}
-
-template <>
-inline void fromUTF8<signed char>(const std::string & src, std::basic_string<signed char> & dest, const std::locale& locale) {
-    dest = fromUTF8<signed char>(src, locale);
-}
-
-template <>
-inline void fromUTF8<unsigned char>(const std::string & src, std::basic_string<unsigned char> & dest, const std::locale& locale) {
-    dest = fromUTF8<unsigned char>(src, locale);
+inline decltype(auto) fromUTF8(const std::string & src) {
+    UnicodeConversionContext context;
+    return fromUTF8<CharType>(src, context);
 }
 
 template <typename CharType>
-inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest,
-    typename std::enable_if<std::is_assignable<
-        std::basic_string<CharType>, decltype(fromUTF8<CharType>(src))
-    >::value>::type * = nullptr
+inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest, UnicodeConversionContext & context,
+    typename std::enable_if_t<
+        std::is_assignable_v<std::basic_string<CharType>, decltype(fromUTF8<CharType>(src))>
+    > * = nullptr
 ) {
-    dest = fromUTF8<CharType>(src);
+    dest = fromUTF8<CharType>(src, context);
 }
 
 template <typename CharType>
-inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest,
-    typename std::enable_if<!std::is_assignable<
-        std::basic_string<CharType>, decltype(fromUTF8<CharType>(src))
-    >::value>::type * = nullptr
+inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest, UnicodeConversionContext & context,
+    typename std::enable_if_t<
+        !std::is_assignable_v<std::basic_string<CharType>, decltype(fromUTF8<CharType>(src))> &&
```

Comment:
same thing as [here](https://github.com/ClickHouse/clickhouse-odbc/pull/280/files#r394941341), but twice...

Could you please add comments on overloads with SFINAE on what is the purpose of specific overload?

Reply:
This is minimally intrusive change to achieve the goal. In a separate PR I am working on changing all this, lots of things that you pointed out are changed already. We just need to release this first, until the main array of changes is ready.

See https://github.com/ClickHouse/clickhouse-odbc/pull/278